pkg/resource/manager.go

```diff
@@ -125,12 +144,245 @@ func (m *manager) BatchCreateGenericResources(ctx context.Context, request *pb.B
+	// the one generated by the filter, if provided
+	queries := filtering.Must{
+		&filtering.Query{
+			Term: &filtering.Term{
+				genericResourceDocumentJoinField: genericResourceRelationName,
+			},
+		},
+	}
+
+	if request.Filter != "" {
+		filterQuery, err := m.filterer.ParseExpression(request.Filter)
+		if err != nil {
+			return nil, err
+		}
+
+		queries = append(queries, filterQuery)
+	}
+
+	searchRequest := &esutil.SearchRequest{
+		Index: m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+		Search: &esutil.EsSearch{
+			Query: &filtering.Query{
+				Bool: &filtering.Bool{
+					Must: &queries,
+				},
+			},
+		},
+	}
+
+	if request.PageSize != 0 {
+		searchRequest.Pagination = &esutil.SearchPaginationOptions{
+			Size:  int(request.PageSize),
+			Token: request.PageToken,
+		}
+	}
+
+	searchResponse, err := m.esClient.Search(ctx, searchRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	var genericResources []*pb.GenericResource
+	for _, hit := range searchResponse.Hits.Hits {
+		var genericResource pb.GenericResource
+		err = protojson.UnmarshalOptions{
+			DiscardUnknown: true,
+		}.Unmarshal(hit.Source, &genericResource)
+		if err != nil {
+			return nil, err
+		}
+
+		genericResource.Id = hit.ID
+		genericResources = append(genericResources, &genericResource)
+	}
+
+	return &pb.ListGenericResourcesResponse{
+		GenericResources: genericResources,
+		NextPageToken:    searchResponse.NextPageToken,
+	}, nil
+}
+
+// ListGenericResourceVersions handles the main logic for fetching versions associated with a generic resource.
+func (m *manager) ListGenericResourceVersions(ctx context.Context, request *pb.ListGenericResourceVersionsRequest) (*pb.ListGenericResourceVersionsResponse, error) {
+	return nil, nil
+}
+
+func (m *manager) GetGenericResource(ctx context.Context, resourceId string) (*pb.GenericResource, error) {
+	log := m.logger.Named("GetGenericResource").With(zap.String("resource", resourceId))
+
+	response, err := m.esClient.Get(ctx, &esutil.GetRequest{
+		Index:      m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+		DocumentId: resourceId,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if !response.Found {
+		log.Debug("generic resource not found")
+
+		return nil, nil
+	}
+
+	var resource pb.GenericResource
+	err = protojson.Unmarshal(response.Source, &resource)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resource, nil
+}
+
+// genericResourceVersionsFromOccurrence will create a map of generic versions, keyed by their IDs, from an occurrence.
+// if the occurrence is a build occurrence, generic resource versions will also be created for each built artifact.
+func genericResourceVersionsFromOccurrence(o *grafeas_proto.Occurrence) map[string]*pb.GenericResourceVersion {
+	result := make(map[string]*pb.GenericResourceVersion)
+
+	result[o.Resource.Uri] = &pb.GenericResourceVersion{
+		Version: o.Resource.Uri,
+	}
+
+	if o.Kind == common_go_proto.NoteKind_BUILD {
```

Comment:
I think this could be done later, but we probably need to treat build occurrences differently in `BatchCreateGenericResources` as well. 

Right now sending a build occurrence will only create a Git generic resource, but any artifacts will produce versions tied to a parent that doesn't exist.

Reply:
Good point.  I might try to get this done in this PR, or I'll open an issue for it.

Reply:
Created https://github.com/rode/rode/issues/83